proxy/src/http/project.rs

```diff
@@ -0,0 +1,422 @@
+    path!("projects" / "register")
+        .and(path::end())
+        .and(warp::post())
+        .and(super::with_registry(registry))
+        .and(warp::body::json())
+        .and_then(handler::register)
+}
+
+/// Project handlers to implement conversion and translation between core domain and http request
+/// fullfilment.
+mod handler {
+    use librad::paths;
+    use librad::surf;
+    use radicle_registry_client::{ed25519, Balance};
+    use std::convert::Infallible;
+    use std::sync::Arc;
+    use tokio::sync::RwLock;
+    use warp::http::StatusCode;
+    use warp::{reply, Rejection, Reply};
+
+    use crate::coco;
+    use crate::project;
+    use crate::registry;
+
+    /// Create a new [`project::Project`].
+    pub async fn create(
+        paths: paths::Paths,
+        input: super::CreateInput,
+    ) -> Result<impl Reply, Rejection> {
+        if surf::git::git2::Repository::open(input.path.clone()).is_err() {
+            coco::init_repo(input.path.clone())?;
+        };
+
+        let (id, meta) = coco::init_project(
+            &paths,
+            &input.path,
+            &input.metadata.name,
+            &input.metadata.description,
+            &input.metadata.default_branch,
+            &input.metadata.img_url,
+        )?;
+
+        Ok(reply::with_status(
+            reply::json(&project::Project {
+                id: librad::project::ProjectId::from(id),
+                metadata: meta.into(),
+            }),
+            StatusCode::CREATED,
+        ))
+    }
+
+    /// Get the [`project::Project`] for the given `id`.
+    pub async fn get(id: String, paths: paths::Paths) -> Result<impl Reply, Rejection> {
+        Ok(reply::json(&project::get(&paths, id.as_ref()).await?))
+    }
+
+    /// List all known projects.
+    pub async fn list(paths: paths::Paths) -> Result<impl Reply, Infallible> {
+        let projects = coco::list_projects(&paths)
+            .into_iter()
+            .map(|(id, meta)| project::Project {
+                id,
+                metadata: meta.into(),
+            })
+            .collect::<Vec<project::Project>>();
+
+        Ok(reply::json(&projects))
+    }
+
+    /// Register a project on the Registry.
+    pub async fn register(
+        registry: Arc<RwLock<registry::Registry>>,
+        input: super::RegisterInput,
+    ) -> Result<impl Reply, Rejection> {
+        // TODO(xla): Get keypair from persistent storage.
+        let fake_pair = radicle_registry_client::ed25519::Pair::from_legacy_string("//Alice", None);
+        // TODO(xla): Use real fee defined by the user.
+        let fake_fee: Balance = 100;
+
+        let mut reg = registry.write().await;
+        let tx = reg
+            .register_project(&fake_pair, input.project_name, input.org_id, None, fake_fee)
+            .await?;
+
+        Ok(reply::with_status(reply::json(&tx), StatusCode::CREATED))
+    }
+}
+
+impl Serialize for project::Project {
```

Comment:
why is Serialize impl'ed here instead of `crate::project` ?

Reply:
I'm on the fence, on one side it's nice to use the derive macros. On the other hand there will be slight diversion between the core types and how they are represented over something like a JSON API. And that is really a concern of the API layer for us `http`. Happy to hear arguments in the other direction.